From 141bb90c13e42fce82bed1e1aefc7f420affd602 Mon Sep 17 00:00:00 2001
From: Hanjie Lin <hanjie.lin@amlogic.com>
Date: Thu, 25 May 2023 16:59:40 +0800
Subject: [PATCH] arm: disable OPTPROBES and noinline handle_pte_fault [1/1]

PD#SWPL-121806

Problem:
arm OPTPROBES will cause kprobe handler unexpected behavior
memory atrace need kprobe at handle_pte_fault()

Solution:
disable HAVE_OPTPROBES and noinline handle_pte_fault

Verify:
sc2

Change-Id: I4efbec5aeeb7b6fe56f6fbd0d413bae123549c5a
Signed-off-by: Hanjie Lin <hanjie.lin@amlogic.com>
---
 arch/arm/Kconfig | 2 +-
 mm/memory.c      | 4 ++++
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index d5bf16462bdb..1d66f014db05 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -121,7 +121,7 @@ config ARM
 	select HAVE_LD_DEAD_CODE_DATA_ELIMINATION if (LD_VERSION >= 23600 || LD_IS_LLD) && LD_CAN_USE_KEEP_IN_OVERLAY
 	select HAVE_MOD_ARCH_SPECIFIC
 	select HAVE_NMI
-	select HAVE_OPTPROBES if !THUMB2_KERNEL
+	select HAVE_OPTPROBES if !THUMB2_KERNEL && (AMLOGIC_DEBUG != y && AMLOGIC_DEBUG != m)
 	select HAVE_PAGE_SIZE_4KB
 	select HAVE_PCI if MMU
 	select HAVE_PERF_EVENTS
diff --git a/mm/memory.c b/mm/memory.c
index 8db0f2a1a4a7..1fc74374d805 100644
--- a/mm/memory.c
+++ b/mm/memory.c
@@ -5787,7 +5787,11 @@ static vm_fault_t wp_huge_pud(struct vm_fault *vmf, pud_t orig_pud)
  * The mmap_lock may have been released depending on flags and our return value.
  * See filemap_fault() and __folio_lock_or_retry().
  */
+#if (defined CONFIG_ARM) && IS_ENABLED(CONFIG_AMLOGIC_DEBUG_ATRACE)
+vm_fault_t handle_pte_fault(struct vm_fault *vmf)
+#else
 static vm_fault_t handle_pte_fault(struct vm_fault *vmf)
+#endif
 {
 	pte_t entry;
 
-- 
2.50.1

